subsys/net/lib/downloader/src/dl_socket.c

```diff
@@ -0,0 +1,400 @@
+
+static const char *str_family(int family)
+{
+	switch (family) {
+	case AF_UNSPEC:
+		return "Unspec";
+	case AF_INET:
+		return "IPv4";
+	case AF_INET6:
+		return "IPv6";
+	default:
+		__ASSERT(false, "Unsupported family");
+		return "Unknown";
+	}
+}
+
+static int socket_sectag_set(int fd, const int *const sec_tag_list, uint8_t sec_tag_count)
+{
+	int err;
+	int verify;
+
+	verify = TLS_PEER_VERIFY_REQUIRED;
+
+	err = setsockopt(fd, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		LOG_ERR("Failed to setup peer verification, errno %d", errno);
+		return -errno;
+	}
+
+	LOG_INF("Setting up TLS credentials, sec tag count %u", sec_tag_count);
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list,
+			 sizeof(sec_tag_t) * sec_tag_count);
+	if (err) {
+		LOG_ERR("Failed to setup socket security tag list, errno %d", errno);
+		return -errno;
+	}
+
+	return 0;
+}
+
+static int socket_tls_hostname_set(int fd, const char *const hostname)
+{
+	__ASSERT_NO_MSG(hostname);
+
+	int err;
+
+	err = setsockopt(fd, SOL_TLS, TLS_HOSTNAME, hostname, strlen(hostname));
+	if (err) {
+		LOG_ERR("Failed to setup TLS hostname (%s), errno %d", hostname, errno);
+		return -errno;
+	}
+
+	return 0;
+}
+
+static int socket_pdn_id_set(int fd, uint32_t pdn_id)
+{
+	int err;
+
+	LOG_INF("Binding to PDN ID: %d", pdn_id);
+	err = setsockopt(fd, SOL_SOCKET, SO_BINDTOPDN, &pdn_id, sizeof(pdn_id));
+	if (err) {
+		LOG_ERR("Failed to bind socket to PDN ID %d, err %d", pdn_id, errno);
+		return -ENETDOWN;
+	}
+
+	return 0;
+}
+
+static int socket_dtls_cid_enable(int fd)
+{
+	int err;
+	uint32_t dtls_cid = TLS_DTLS_CID_ENABLED;
+
+	err = setsockopt(fd, SOL_TLS, TLS_DTLS_CID, &dtls_cid, sizeof(dtls_cid));
+	if (err) {
+		err = -errno;
+		LOG_ERR("Failed to enable TLS_DTLS_CID: %d", err);
+		/* Not fatal, so continue */
+	}
+
+	return err;
+}
+
+static bool is_ip_address(const char *hostname)
+{
+	struct sockaddr sa;
+
+	if (zsock_inet_pton(AF_INET, hostname, sa.data) == 1) {
+		return true;
+	} else if (zsock_inet_pton(AF_INET6, hostname, sa.data) == 1) {
+		return true;
+	}
+
+	return false;
+}
+
+static int dl_socket_host_lookup(const char *const hostname, uint32_t pdn_id, struct sockaddr *sa,
+				 int family)
+{
+	int err;
+	char pdnserv[4];
+	char *servname = NULL;
+	struct addrinfo *ai;
+	struct addrinfo hints = {
+		.ai_family = family,
+	};
+
+#if !defined(CONFIG_NET_IPV6)
+	if (family == AF_INET6) {
+		return -EINVAL;
+	}
+#endif
+
+#if !defined(CONFIG_NET_IPV4)
+	if (family == AF_INET) {
+		return -EINVAL;
+	}
+#endif
+
+	LOG_DBG("host lookup %s, pdn id %d, family %d\n", hostname, pdn_id, family);
+
+	if (pdn_id) {
+		hints.ai_flags = AI_PDNSERV;
+		(void)snprintf(pdnserv, sizeof(pdnserv), "%d", pdn_id);
+		servname = pdnserv;
+	}
+
+	err = getaddrinfo(hostname, servname, &hints, &ai);
+	if (err) {
+		/* We expect this to fail on IPv6 sometimes */
+		LOG_DBG("Failed to resolve hostname %s on %s, err %d\n", hostname,
```

Comment:
Same as above, `\n` not needed. Also, personally I liked having this as a `LOG_WRN` to know that a fallback might happen.

Reply:
> Also, personally I liked having this as a `LOG_WRN` to know that a fallback might happen (or not).

Hmmm, the caller doesn't seem to trigger fallback when this failure happens but `dl_socket_configure_and_connect()` simply fails. An error is logged from the caller anyway. Hence, it doesn't matter much whether this is `LOG_DBG`, `LOG_WRN` or `LOG_ERR` but if this is fatal error that will give an error to the downloader API user, this should be WRN or ERR. That being said, we don't need duplicate logs from here and the caller. I'd keep the log here, change it to LOG_WRN, and remove the logging from the caller.

Reply:
Updated.

Reply:
> the caller doesn't seem to trigger fallback when this failure happens but dl_socket_configure_and_connect() simply fails

I'm not sure this is the behaviour I'm seeing. If `dl_host_cfg->family` is not specified, then the fallback happens automatically in `dl_socket_configure_and_connect()`, without further feedback to the caller of the API (only the DBG log I highlighted above).

Reply:
@kacperradoszewski You are right, the fallback happens in `dl_socket_configure_and_connect()` unless the family is explicitly set (`AF_INET` or `AF_INET6`). In the case the fallback occurs, the user will now get a warning that we fail to connect with IPv6. If both fail, you get two warning logs, one for IPv6 and one for IPv4. Previously you would get a debug log for each failed attempt and a generic log error that you fail to connect. Perhaps that is more clear to the user in the end?

Reply:
Anyway, I might have imagined that this has ever been a WRN now that I checked the current implementation, so feel free to disregard and keep as is.

Reply:
> @kacperradoszewski You are right, the fallback happens in `dl_socket_configure_and_connect()` unless the family is explicitly set (`AF_INET` or `AF_INET6`).

True.

> In the case the fallback occurs, the user will now get a warning that we fail to connect with IPv6. If both fail, you get two warning logs, one for IPv6 and one for IPv4. Previously you would get a debug log for each failed attempt and a generic log error that you fail to connect. Perhaps that is more clear to the user in the end?

> Though the debug log can also be `LOG_WRN` to highlight that we are using the fallback.

`LOG_WRN` is always a bit problematic as user will start to look at what he is doing wrong, which may result in customer support ticket too.
Although I use `LOG_INF` rarely, because it's default log level and you don't want debug logging on that level, this is an example where I'd use it. Unless this warning log is a suggestion to set the IP family by the user of the API, in which case LOG_WRN is good.

Reply:
> Although I use LOG_INF rarely, because it's default log level and you don't want debug logging on that level, this is an example where I'd use it.

That sounds good to me.

Reply:
I see your point with `LOG_WRN` being problematic. The fallback must expected to be used in some cases I think, so I will update the log level.

Reply:
I'm fine with this although I'm starting to think why we don't also clearly log (maybe with `LOG_INF`) that DNS fallback is occurring in an else-branch on line 285. Getting clear information on what's going on might save some engineers time and maybe even a support ticket in some case.

Reply:
Added `LOG_INF` on DNS fallback.